opentype/api/font/cff/interpreter/interpreter.go

```diff
@@ -0,0 +1,363 @@
+		p.instructions = p.instructions[5:]
+	}
+
+	if hasResult {
+		if p.ArgStack.Top == psArgStackSize {
+			return true, errInvalidCFFTable
+		}
+		p.ArgStack.Vals[p.ArgStack.Top] = number
+		p.ArgStack.Top++
+	}
+	return hasResult, nil
+}
+
+const maxNibbleDefsLength = len("E-")
+
+// nibbleDefs encodes 5176.CFF.pdf Table 5 "Nibble Definitions".
+var nibbleDefs = [16]string{
+	0x00: "0",
+	0x01: "1",
+	0x02: "2",
+	0x03: "3",
+	0x04: "4",
+	0x05: "5",
+	0x06: "6",
+	0x07: "7",
+	0x08: "8",
+	0x09: "9",
+	0x0a: ".",
+	0x0b: "E",
+	0x0c: "E-",
+	0x0d: "",
+	0x0e: "-",
+	0x0f: "",
+}
+
+// subrBias returns the subroutine index bias as per 5177.Type2.pdf section 4.7
+// "Subroutine Operators".
+func subrBias(numSubroutines int) int32 {
+	if numSubroutines < 1240 {
+		return 107
+	}
+	if numSubroutines < 33900 {
+		return 1131
+	}
+	return 32768
+}
+
+// CallSubroutine calls the subroutine, identified by its index, as found
+// in the instructions (that is, before applying the subroutine biased).
+// `isLocal` controls whether the local or global subroutines are used.
+// No argument stack modification is performed.
+func (p *Machine) CallSubroutine(index int32, isLocal bool) error {
+	subrs := p.globalSubrs
+	if isLocal {
+		subrs = p.localSubrs
+	}
+
+	// no bias in type1 fonts
+	if p.ctx == Type2Charstring {
+		index += subrBias(len(subrs))
+	}
+
+	if index < 0 || int(index) >= len(subrs) {
+		return fmt.Errorf("invalid subroutine index %d (for length %d)", index, len(subrs))
+	}
+	if p.callStack.top == psCallStackSize {
+		return errors.New("maximum call stack size reached")
+	}
+	// save the current instructions
+	p.callStack.vals[p.callStack.top] = p.instructions
+	p.callStack.top++
+
+	// activate the subroutine
+	p.instructions = subrs[index]
+	return nil
+}
+
+// Return returns from a subroutine call.
+func (p *Machine) Return() error {
+	if p.callStack.top <= 0 {
+		return errors.New("no subroutine has been called")
+	}
+	p.callStack.top--
+	// restore the previous instructions
+	p.instructions = p.callStack.vals[p.callStack.top]
+	return nil
+}
+
+// PsOperator is a postcript command, which may be escaped.
+type PsOperator struct {
+	Operator  byte
+	IsEscaped bool
+}
+
+func (p PsOperator) String() string {
+	if p.IsEscaped {
+		return fmt.Sprintf("2-byte operator (12 %d)", p.Operator)
+	}
+	return fmt.Sprintf("1-byte operator (%d)", p.Operator)
+}
+
+// PsOperatorHandler defines the behaviour of an operator.
+type PsOperatorHandler interface {
+	// Context defines the precise behaviour of the interpreter,
+	// which has small nuances depending on the context.
+	Context() PsContext
+
+	// Apply implements the operator defined by `operator` (which is the second byte if `escaped` is true).
+	//
+	// Returning `ErrInterrupt` stop the parsing of the instructions, without reporting an error.
+	// It can be used as an optimization.
+	Apply(operator PsOperator, state *Machine) error
```

Comment:
usually, in Go APIs, it seems to me we try to put the modified argument "first":
```suggestion
	Apply(state *Machine, op Operator) error
```

Reply:
Done.